File: platforms/qt/treeitem.cpp
#include "treeitem.h"

#include <QDebug>

int TreeItem::isExistElementInList(QList<QTreeWidgetItem *> *items, QString keyword) {
    for (int i = 0; i < items->size(); i++) {
        if (items->at(i)->text(default_column) == keyword) {
            return i;
        }
    }
    qDebug() << "WARNING: isExistElement not found";
    return -1;
}

int TreeItem::isExistChildInElement(QTreeWidgetItem *item, QString keyword) {
    for (int i = 0; i < item->childCount(); i++) {
        if (item->child(i)->text(default_column) == keyword) {
            return i;
        }
    }
    return -1;
}

int TreeItem::sortChildrenOfItem(QTreeWidgetItem *item) {
    for (int i = 0; i < item->childCount(); i++) {
        for (int n = i; n >= 0; n--) {
            if (item->child(i)->text(default_column).toInt() < item->child(n)->text(default_column).toInt()) {
                QTreeWidgetItem buff;
                buff = *item->child(i);
                *item->child(i) = *item->child(n);
                *item->child(n) = buff;
                i = n;
            }
        }
    }

    return 0;
}

int TreeItem::appendOnceElement(QTreeWidgetItem *item, QString text) {
    int retval = isExistChildInElement(item, text);
    if (retval == -1) {
        retval = item->childCount();
        QTreeWidgetItem *buffItem = new QTreeWidgetItem(QStringList(text));
        buffItem->setText(default_column, text);
        item->insertChild(retval, buffItem);
    }
    return retval;
}

int TreeItem::appendNewDate(QString year, QString month, QString day) {
    QTreeWidgetItem *buffItem = nullptr;

    int retval = -1;
    retval = isExistElementInList(&yearItems, year);

    if (retval == -1) {
        QTreeWidgetItem *yearItem = new QTreeWidgetItem(QStringList(year));
        yearItem->setText(default_column, year);
        retval = yearItems.size();
        yearItems.append(yearItem);
    }

    buffItem = dynamic_cast<QTreeWidgetItem *> (yearItems.at(retval));

    retval = appendOnceElement(buffItem, month);

    buffItem = buffItem->child(retval);

    retval = appendOnceElement(buffItem, day);

    return 0;
}

int TreeItem::fillTreeByDateList(QList<QDate *> *dateList) {
    Q_ASSERT(dateList->size() >= 1);

    QListIterator<QDate *> iterator(*dateList);

    int year; int month; int day;

    do {
        iterator.next()->getDate(&year, &month, &day);
        qDebug() << "DEBUG: fillTreeByDateTimeList year:" << year << ", month: " << month << ", day: " << day;

        appendNewDate(QString::number(year), QString::number(month), QString::number(day));

    } while(iterator.hasNext());

    treeWidget->insertTopLevelItems(0, yearItems);
    return 0;
}

void TreeItem::getDaysList(QList<QTreeWidgetItem *> *items) {
    QTreeWidgetItem *buffItem;
    for (int i = 0; i < yearItems.size(); i++) {
        if (yearItems.at(i) == nullptr) continue;
        for (int k = 0; k < yearItems.at(i)->childCount(); k++) {
            if (yearItems.at(i)->child(k) == nullptr) continue;
            buffItem = yearItems.at(i)->child(k);
            for (int n = 0; n < buffItem->childCount(); n++) {
                if (buffItem->child(n) == nullptr) continue;
                items->append(buffItem->child(n));
            }
        }
    }
}

int TreeItem::installSignalsForExistedTree(void (*handler)(QTreeWidgetItem *item)) {
    QTreeWidgetItem *monthItems = nullptr;

    for (int i = 0; i < yearItems.size(); i++) {
        monthItems = dynamic_cast<QTreeWidgetItem *> (yearItems.at(i));
        if (monthItems == nullptr) continue;
        for (int k = 0; k < monthItems->childCount(); k++) {
            handler(monthItems->child(k));
        }
    }
    return 0;
}

void TreeItem::clear() {
    QListIterator<QTreeWidgetItem *> iteratorYears(yearItems);

    for (int i = 0; i < yearItems.length(); i++) {
        QTreeWidgetItem *month = iteratorYears.next()->child(i);
        for (int n = 0; n < month->childCount(); n++) {
            month->removeChild(month->child(n));
        }
    }

    yearItems.clear();
}

TreeItem::~TreeItem(){
    this->clear();
}

TreeItem::TreeItem(QTreeWidget *treeWidget, int default_column) {
    this->default_column = default_column;
    this->treeWidget = treeWidget;
    this->treeWidget->setColumnCount(default_column);
}

File: platforms/android/app/src/main/java/TreeItem.java
import android.util.Log;

import com.example.note2.TreeNode;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

public class TreeItem {
    private static final String TAG = "TreeItem";

    private int defaultColumn;
    // For simplicity, we store the top-level items (years) in a list.
    private List<TreeNode> yearItems;
    // If you have a custom tree view widget, you might hold a reference here.
    private Object treeWidget;

    public TreeItem(Object treeWidget, int defaultColumn) {
        this.defaultColumn = defaultColumn;
        this.treeWidget = treeWidget;
        this.yearItems = new ArrayList<>();
        // You might configure your treeWidget here if needed.
    }

    // Searches for an element in a list of TreeNodes that has matching text.
    public int isExistElementInList(List<TreeNode> items, String keyword) {
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).getText().equals(keyword)) {
                return i;
            }
        }
        Log.w(TAG, "WARNING: isExistElement not found");
        return -1;
    }

    // Searches for a child element of a given TreeNode with matching text.
    public int isExistChildInElement(TreeNode item, String keyword) {
        List<TreeNode> children = item.getChildren();
        for (int i = 0; i < children.size(); i++) {
            if (children.get(i).getText().equals(keyword)) {
                return i;
            }
        }
        return -1;
    }

    // Sorts the children of a given TreeNode numerically based on the text.
    public int sortChildrenOfItem(TreeNode item) {
        List<TreeNode> children = item.getChildren();
        Collections.sort(children, new Comparator<TreeNode>() {
            @Override
            public int compare(TreeNode o1, TreeNode o2) {
                try {
                    int i1 = Integer.parseInt(o1.getText());
                    int i2 = Integer.parseInt(o2.getText());
                    return Integer.compare(i1, i2);
                } catch (NumberFormatException e) {
                    return o1.getText().compareTo(o2.getText());
                }
            }
        });
        return 0;
    }

    // Appends a new child element with the given text to a TreeNode, but only if it does not already exist.
    public int appendOnceElement(TreeNode item, String text) {
        int index = isExistChildInElement(item, text);
        if (index == -1) {
            index = item.getChildren().size();
            TreeNode newNode = new TreeNode(text);
            item.getChildren().add(index, newNode);
        }
        return index;
    }

    // Appends a new date to the tree. The date is provided as three strings: year, month, and day.
    public int appendNewDate(String year, String month, String day) {
        int yearIndex = isExistElementInList(yearItems, year);
        if (yearIndex == -1) {
            TreeNode yearNode = new TreeNode(year);
            yearIndex = yearItems.size();
            yearItems.add(yearNode);
        }
        TreeNode yearNode = yearItems.get(yearIndex);
        int monthIndex = appendOnceElement(yearNode, month);
        TreeNode monthNode = yearNode.getChildren().get(monthIndex);
        int dayIndex = appendOnceElement(monthNode, day);
        return 0;
    }

    // Fills the tree using a list of Date objects.
    public int fillTreeByDateList(List<Date> dateList) {
        if (dateList.isEmpty()) {
            throw new AssertionError("dateList must contain at least one element");
        }
        for (Date date : dateList) {
            Calendar cal = Calendar.getInstance();
            cal.setTime(date);
            int year = cal.get(Calendar.YEAR);
            int month = cal.get(Calendar.MONTH) + 1;  // Calendar.MONTH is zero-based.
            int day = cal.get(Calendar.DAY_OF_MONTH);
            Log.d(TAG, "DEBUG: fillTreeByDateList year:" + year + ", month:" + month + ", day:" + day);
            appendNewDate(String.valueOf(year), String.valueOf(month), String.valueOf(day));
        }
        // In Qt, the yearItems were inserted into a QTreeWidget.
        // In Android, update your adapter/view with the 'yearItems' list as needed.
        return 0;
    }

    // Collects all day-level nodes into the provided list.
    public void getDaysList(List<TreeNode> items) {
        for (TreeNode yearNode : yearItems) {
            if (yearNode == null) continue;
            for (TreeNode monthNode : yearNode.getChildren()) {
                if (monthNode == null) continue;
                for (TreeNode dayNode : monthNode.getChildren()) {
                    if (dayNode == null) continue;
                    items.add(dayNode);
                }
            }
        }
    }

    // Installs a signal (callback) on every month node in the tree.
    // The handler is a functional interface that accepts a TreeNode.
    public int installSignalsForExistedTree(Handler handler) {
        for (TreeNode yearNode : yearItems) {
            if (yearNode == null) continue;
            for (TreeNode monthNode : yearNode.getChildren()) {
                handler.onItem(monthNode);
            }
        }
        return 0;
    }

    // Clears the tree.
    public void clear() {
        for (TreeNode yearNode : yearItems) {
            if (yearNode == null) continue;
            for (TreeNode monthNode : yearNode.getChildren()) {
                monthNode.getChildren().clear();
            }
        }
        yearItems.clear();
    }

    // In Java, there is no destructor; call clear() when needed.
    // Optionally, override finalize() (not recommended) or manage cleanup elsewhere.

    // Getter for the year items (could be used to update an adapter).
    public List<TreeNode> getYearItems() {
        return yearItems;
    }

    // A simple interface to mimic the signal/slot handler for items.
    public interface Handler {
        void onItem(TreeNode item);
    }
}
